Contains a number of fixes:

johnlevidy opened the pull request:
1. Rewrites date densifications
2. Better displaying of start / end for parallelizable tasks
3. Display parallelizable tasks better
4. Dont ever show completed items in red
5. Add a check for incomplete ancestors ( this breaks for parallelizable tasks without other changes )

johnlevidy committed:
Contains a number of fixes:

1. Rewrites date densifications
2. Better displaying of start / end for parallelizable tasks
3. Display parallelizable tasks better
4. Dont ever show completed items in red
5. Add a check for incomplete ancestors ( this breaks for parallelizable
   tasks without other changes )

johnlevidy commented on backend_rewrite/scheduler.py:
```diff
@@ -191,31 +191,49 @@ def get_assignees(task: InputTask, metadata: Metadata, person_to_person_id: bidi
-    start_offset = busdays_between(today, start) if start else 0
-    end_offset = busdays_between(today, end) if end else horizon
-    exclude = end_offset < 0
+# TODO: I have a feeling this might still be wrong for paralleliable, 
```
@nyscotsman This has me pretty nervous. I'm not sure the best way to handle this.

nyscotsman replied:
Perhaps this needs to be done before splitting the task. Probably for the chain of 1 day tasks it's only the earliest start of the first day and the latest end of the last that is relevant; the edges between the days keeps the ordering within the task.

johnlevidy replied:
I think the easiest way to do it is to modify build_graph_and_schedule to own the "today" rollback, then to pass "today" into the task expansion. I think it also lets us get rid of the special case in the densification, you'd always have a linear chain of assumed incomplete tasks in future and a blob of completed time in the past

nyscotsman commented on backend_rewrite/verify.py:
```diff
@@ -16,19 +16,32 @@ def find_cycle(G: nx.Graph) -> Optional[Any]:
-        if not task.start_date or not task.end_date:
-            continue
-        if task.start_date >= task.end_date and task.estimate > 0:
+        if task.start_date and task.end_date and task.start_date >= task.end_date and task.estimate > 0:
```
Does this need to be broken up a bit? If estimate is set and zero, we should probably require start == end. If estimate is not set or not zero, start must be less than end.

johnlevidy committed:
Number of fixes

1. Adds step param to the scheduler to make testing a bit sturdier
2. Pulls rollback up a level so that task expansion can be
   rollback-aware
3. Task expansion now handles parallelizable tasks correctly -- where
   the task is scheduled in a way that puts as much remaining time on
   the task in the future.
4. Other date densification is simplified in light of (3)
